File: Sponsors/mw-streams-java/src/main/java/com/mwam/hackathon/operators/FlatMap.java
package com.mwam.hackathon.operators;

import java.util.ArrayList;
import java.util.Collection;
import java.util.function.Function;

public class FlatMap<I, O> extends Operator<I, O> {

    private final Function<I, Collection<O>> fn;

    public FlatMap(Function<I, Collection<O>> fn) {
        super(new ArrayList<>());
        this.fn = fn;
    }

    /**
     * Consumes an input and maps it to a Collection.
     * Emits each value of the collection as a separate event.
     *
     * @param input to be flattened.
     */
    @Override
    public void consume(I input) {
        final Collection<O> outputCollection = fn.apply(input);
        outputCollection.forEach(this::emit);
    }
}

File: Sponsors/mw_streams_python/mws/operators/flatmap.py
from typing import Iterable, TypeVar, Callable

from mws.operators.operator import Operator


I = TypeVar("I")
O = TypeVar("O")


class FlatMap(Operator[I, O]):
    def __init__(self, fn: Callable[[I], Iterable[O]]) -> None:
        super().__init__()
        self.fn = fn

    def consume(self, event: I) -> None:
        raise NotImplementedError("TODO implement me")
